File: 21.03.10/Programmers_72411.py
from itertools import combinations

# def check(orders, cnt, menu_dic):
#     for order in orders:
#         order = sorted(list(map(str, order)))
#         lst = list(combinations(order, cnt))
#         for val in lst:
#             tmp = "".join(val)
#             if menu_dic.get(tmp):
#                 menu_dic[tmp] += 1
#             else:
#                 menu_dic[tmp] = 1
#     return menu_dic


# def solution(orders, course):
#     answer = []

#     for cnt in course:
#         menu_dic = {}
#         menu_dic = check(orders, cnt, menu_dic)

#         maxi = 0
#         menu = []
#         for key, val in menu_dic.items():
#             if val > 1:
#                 if val > maxi:
#                     maxi = val
#                     menu = [key]
#                 elif val == maxi:
#                     menu.append(key)
#         if menu:
#             answer += menu
    
#     answer.sort()

#     return answer



# use Counter

from collections import Counter

def solution(orders, course):
    answer = []

    for course_size in course:
        order_comb = []
        for order in orders:
            order_comb += combinations(sorted(order), course_size)

        most_ordered = Counter(order_comb).most_common()
        for menu, cnt in most_ordered:
            if cnt > 1 and cnt == most_ordered[0][1]:
                answer.append("".join(menu))

    answer.sort()

    return answer




print(solution(["ABCFG", "AC", "CDE", "ACDE", "BCFG", "ACDEH"], [2, 3, 4]))
print(solution(["ABCDE", "AB", "CD", "ADE", "XYZ", "XYZ", "ACD"], [2, 3, 5]))
print(solution(["XYZ", "XWY", "WXA"], [2, 3, 4]))

# answer
# ["AC", "ACDE", "BCFG", "CDE"]
# ["ACD", "AD", "ADE", "CD", "XYZ"]
# ["WX", "XY"]

File: 21.04.19/Programmers_72411.js
function combination(answer, source, target, n, r, count) {
  if (r === 0) answer.push(target);
  else if (n === 0 || n < r) return;
  else {
    target += source[count];
    combination(source, Object.assign('', target), n - 1, r - 1, count + 1);
    target = target.slice(0, -1);
    combination(source, Object.assign('', target), n - 1, r, count + 1);
  }
}


function solution(orders, course) {
  var answer = [];

  const orderedCount = new Map();
  for (let i = 0; i < course.length; i++) {
    combination(answer, orders, '', orders.length, course[i], 0);
  }
  
  answer.sort();

  return answer;
}




console.log(solution(["ABCFG", "AC", "CDE", "ACDE", "BCFG", "ACDEH"], [2,3,4]))
console.log(solution(["ABCDE", "AB", "CD", "ADE", "XYZ", "XYZ", "ACD"], [2,3,5]))
console.log(solution(["XYZ", "XWY", "WXA"], [2,3,4]))

// answer
// ["AC", "ACDE", "BCFG", "CDE"]
// ["ACD", "AD", "ADE", "CD", "XYZ"]
// ["WX", "XY"]